Added vanadium, absorption correction to workflow

celinedurniak opened the pull request.

celinedurniak committed:
Added vanadium, absorption correction to workflow

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+              radius: radius of cylindrical sample in centimeters
+
+              scattering: coherent+incoherent cross section at 1.8 Angstroms
+                          in barns
```
Why not use the sample information that comes with the file?

celinedurniak replied:
I cropped the details about **absorp to

**absorp: dictionary containing information to correct absorption  
see help of Mantid's algorithm CylinderAbsorption for details   https://docs.mantidproject.org/nightly/algorithms/CylinderAbsorption-v1.html

celinedurniak replied:
Which file are you talking about?

SimonHeybrock replied:
The NeXus files.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+                                           ScatteringXSection=absorp['scattering'],
+                                           AttenuationXSection=absorp['attenuation'],
+                                           CylinderSampleHeight=absorp['height'],
+                                           CylinderSampleRadius=absorp['radius'])
```
Is this required? If the files contain a sample consider removing it to keep things simple?

celinedurniak replied:
The minimum number of arguments is sample, lambda_binning, CylinderSampleHeight and CylinderSampleRadius.

SimonHeybrock replied:
Shouldn't the file contain the sample information such as height and radius?

celinedurniak replied:
I just loaded the Nexus file for the sample, checked the sample logs and sample material. There is no info about the height, radius of the sample shape.

![Screen Shot 2020-03-23 at 14 15 42](https://user-images.githubusercontent.com/7889724/77320421-d5acd700-6d10-11ea-86e5-5780d9c959f8.png)

SimonHeybrock replied:
Does the WISH workflow run absorption correction for the sample? Maybe just for vanadium?

celinedurniak replied:
I just checked reduce.py. I think the absorption is also applied to the sample data.
I also loaded the Vanadium nexus file into Mantid. Same thing: no info about the sample in Sample material.

SimonHeybrock replied:
Surprising (how do the users know the sizes for any given sample?), but I suppose we have to leave it in then. 👍

celinedurniak replied:
I cropped the input arguments to the minimum required

SimonHeybrock replied:
Have you considered using a `dict`, as scipp does with the `mantid_args` for `LoadEventNexus`?

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    # proceed to the normalization
+
+    correction.attrs['source_position'] = sample_lambda.attrs['source_position'].copy()
+    correction.attrs['sample_position'] = sample_lambda.attrs['sample_position'].copy()
```
Not needed?

celinedurniak replied:
I just checked. These 2 lines are not needed anymore. They'll be removed.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    # 4. Focus panels
     # Assuming sample is in d-spacing: Focus into groups
+    # deleting info in sample_dspacing as temporary fix bacause of bug in scipp
+
```
What does this comment refer to?

celinedurniak replied:
I removed the line of code but forgot to remove the comment above it.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+                                             neginf=replace_value)
+
+        focused_hist['group', i].values = values_no_nan_inf.values
+        focused_hist['group', i].variances = variances_no_nan_inf.values
```
Why are you doing this group by group?

Furthermore, I think setting values and variances to 0 should lead to bad results. Shouldn't bad values be masked instead?

celinedurniak replied:
This part of the script has been removed in order to get events as output.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    # Sum final result for plotting
+    sample_sum = sc.sum(focused_hist, 'group')
+
+    return sample_sum
```
Return event data, so the user of the function can decide whether they want to histogram or not?

celinedurniak replied:
So you would remove from 6.). Am I correct?

SimonHeybrock replied:
I think so. I suppose you can move the code to an independent helper function for convenience, but not call it here.

celinedurniak replied:
OK. So one of the points (replacement of Nan, inf, -inf) is not relevant anymore.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    # convert to wavelength
+    mon4_lambda = sc.neutron.convert(mon4_selected, 'tof', 'wavelength')
+
+    del mon4_selected
```
Shorter?
```suggestion
    mon4_lambda = sc.neutron.convert(event_data.attrs['monitor4'].value, 'tof', 'wavelength')
```

celinedurniak replied:
OK I'll merge the 2 lines here and for the Sample data as well

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    # del vana_mon4_lambda
+    # smooth data
+    # vana_mon4_smooth = smooth_data(vana_mon4_spline, dim='wavelength', NPoints=40)
+    # del vana_mon4_spline
```
Commented code?

celinedurniak replied:
I'll remove these lines

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+                                           AttenuationXSection=absorp['attenuation'],
+                                           CylinderSampleHeight=absorp['height'],
+                                           CylinderSampleRadius=absorp['radius'])
+    else:
```
See above.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    # proceed to the normalization
+
+    correction.attrs['source_position'] = vana_red.attrs['source_position'].copy()
+    correction.attrs['sample_position'] = vana_red.attrs['sample_position'].copy()
```
Not necessary?

celinedurniak replied:
Will be removed.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
-    del focused
+    # Calibration
+    # Load
+    input_load_cal = {'InstrumentFilename': 'WISH_Definition.xml'}
+    calvana = load_calibration(calibration, mantid_args=input_load_cal)
```
Does this work properly now? When I ran this last week Mantid was really slow and had apparently trouble with tons of bad detector IDs.

celinedurniak replied:
I had the same issues: slow and message about bad detectors. But it is the only way I found to get the groups for the correct amount of spectrum. 
This issue is due to the fact that Sample and Vanadium data were not recorded using the same layout of detectors. In normal cases, it should not be the case.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+        single_spectrum = vana_rebin['spectrum', i].values
+        detect_if_negative_values = any(item < 0 for item in single_spectrum)
+        if detect_if_negative_values:
+            vana_rebin.masks['mask'].values[i] = True
```
So you are masking the complete spectrum even if there is just a single negative bin? Why not masks bin-by-bin?

```python
vana_rebin.masks['negative_bins'] = sc.Variable(dims=..., values=np.less(vana_rebin.values, 0.0))
```
```

celinedurniak replied:
I modified the script accordingly.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+        if detect_if_negative_values:
+            vana_rebin.masks['mask'].values[i] = True
+
+    # focus (deleting info in vana_dspacing because of bug in scipp)
```
What is this referring to?

celinedurniak replied:
You mean line 463 only?

SimonHeybrock replied:
Yes.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    # del vana_dspacing
+
+    # #smooth_data
+    # smooth_data(vana_focus_spline, dim='d-spacing', NPoints=40)
```
Commented code?

celinedurniak replied:
These steps are in the initial WISH workflow. But since I had issues with applying spline_background, I commented these 2 lines out.

SimonHeybrock replied:
This is not clear to the reader. Commented code should be avoided since it makes the purpose unclear:
- Is there a bug?
- Has this been accidentially disabled?
- Is it broken?
- Is it still needed?
- Harder to read?
- `git` remembers code, why not remove it.

celinedurniak replied:
OK I'll remove these lines here and also from line 283.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,459 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    #                                 calibration="WISH_cycle_15_4_noends_10to10_dodgytube_removed_feb2016.cal",
+    #                                 vanadium=None,
+    #                                 empty_instr=None,
+    #                                 lambda_binning=(0.7, 10.35, 5615))
```
Commented code?

celinedurniak replied:
It was the old implementation without absorption. I'll remove this part.

SimonHeybrock commented on spline_background.py:
```diff
@@ -32,57 +35,45 @@ def bspline_background(variable, dim, smoothing_factor=None):
+
+        new_x_values = [0.5*(x_values[i] + x_values[i+1])
+                        for i in range(len(x_values)-1)]
+        x_values = new_x_values
```
```suggestion
        x_values = 0.5*(x_values[:-1] + x_values[1:])
```

celinedurniak replied:
I modify the script to this new version

SimonHeybrock commented on spline_background.py:
```diff
@@ -32,57 +35,45 @@ def bspline_background(variable, dim, smoothing_factor=None):
+        output_x = sc.Variable(dims=[dim], values=bin_edge_output_x)
+    else:
+        output_x = sc.Variable(dims=[dim], values=splined[0])
+
```
```suggestion
    output_x = variable.coords[dim]
```

celinedurniak replied:
The input x-values cannot be simply copied to the output because the scipy algorithms modify it.

SimonHeybrock replied:
Ah, ok. You can still shorten by doing parts of the math with numpy arrays though.

SimonHeybrock commented on spline_background.py:
```diff
@@ -32,57 +35,45 @@ def bspline_background(variable, dim, smoothing_factor=None):
     weights = 1.0/errors
 
-    # https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.UnivariateSpline.html
+    # flag marking 'bin-edge' status of input x values
+    bin_edge_input = False
```
Not needed anymore (see common on using input X as output X).

celinedurniak committed:
Update powder_reduction.py

celinedurniak committed:
Update spline_background.py

celinedurniak committed:
Update spline_background.py

celinedurniak committed:
Update spline_background.py

celinedurniak committed:
Update spline_background.py

Modified method to calculate final bin-edges axis

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -23,111 +44,346 @@ def powder_reduction(sample='sample.nxs', calibration=None, lambda_binning=(0.7,
+    del mon_rebin, mon4_smooth
+
+    # 2. absorption correction
+    # TODO: adapt for different number of inputs
```
Why not just unpack the dict (dict keys have to match Mantid argument names) below:
```
bsorption_correction(sample, lambda_binning, **absorp)
```
Then there is no "TODO" and it should be flexible and keep working irrespective of future changes?

SimonHeybrock replied:
Or maybe even better: Just pass the dict to `absorption_correction` and let that do the unpacking?

celinedurniak replied:
I just submitted a new version of the file to deal with different values for the absorption correction related to Sample and Vanadium.

SimonHeybrock commented on powder_reduction.py:
```diff
@@ -3,14 +3,35 @@
+                     empty_instr=None,
+                     lambda_binning=(0.7, 10.35, 5615),
+                     dspacing_binning=(1., 10., 9000),
+                     **absorp):
```
This implies that the same sample size is used for sample and vanadium runs. Is guess this is not true in practice, i.e., you should probably have independent absorption parameters for the two?

celinedurniak committed:
Update powder_reduction.py

Used different input values related to absorption correction for Sample and Vanadium
